proposal: Talent Plan Courses - Distributed Database System (#129)
pingcap/community: rfc/2019-12-9-pingcap-university-campus.md

# Proposal: Talent Plan Courses
- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2019-12-6
- Discussion at:

## Abstract

There are plenty of people interested in TiDB but lack enough background knowledge, especially college students and engineers without backend development experience. The proposal aims at helping those people enter the field of TiDB by creating a series of courses, which is also called Talent Plan Courses. Participants should be able to pull requests to TiDB related projects after consumed the course materials.

## Background

> Since the campus is always an important member of the TiDB Community, and we hope more young blood from universities or other fields can embrace the TiDB ecosystem, so we would develop a distributed database system course to help people without enough background knowledge to enter the field. The courses will be consist of reading supplies and programming assignments. As there are already many accessible materials online, we can develop the content of the course by learning from them. We think the voice from the different populations is important for the success of the goal, especially college students, so we hope they can join us to make this happen.

Reviewer: Consider combining these two sections. This is not a thesis paper, no abstract is needed. Keep it simple and short. Don't have to repeat the same thing twice.

## Proposal

> The course is consists of two parts, reading materials and programming assignments. The reading materials can be collected from textbooks, slides, and notes from other related courses.

Reviewer: ```suggestion
The course consists of two parts - reading materials and programming assignments. 
```

### Reading Materials

> The list of reference for reading materials, both for database system and distributed system:

Reviewer: ```suggestion
Reading materials can be collected from textbooks, slides, and notes from other related courses. The following is a  list of reading materials, covering database systems and distributed systems:
```

- [Reading in Database system, 5th Edition](http://www.redbook.io/)
- [Architecture of a Database system](http://db.cs.berkeley.edu/papers/fntdb07-architecture.pdf)
- [Database System Concept](https://www.db-book.com/db6/index.html)
- [Database System Implementation](http://infolab.stanford.edu/~ullman/dbsi.html)
- [CMU cs15445 Database System](https://15445.courses.cs.cmu.edu/fall2019/schedule.html)
- [CMU cs15721 Advanced Database System](https://15721.courses.cs.cmu.edu/spring2019/schedule.html)
- [UCB cs186 course materials](https://cs186berkeley.net/)
- [MIT 6.830 course materials](http://db.csail.mit.edu/6.830/)
- [Stanford cs245 course materials](http://web.stanford.edu/class/cs245/)
- [MIT 6.824 Distributed Systems](https://pdos.csail.mit.edu/6.824/)
- [Princeton COS418 Distributed Systems](https://www.cs.princeton.edu/courses/archive/fall19/cos418/)
- [Princeton COS316 Principle of Computer System Design](https://www.cs.princeton.edu/courses/archive/fall19/cos316/)
- [UCB Operating Systems and System Programming](https://cs162.eecs.berkeley.edu/)
- [Distributed System 3rd Edition](https://www.distributed-systems.net/)

> Many important topics have already been covered in the above. Course development participants are required to reorganize the content into slides or notes and translate them into Chinese, We hope the final content should be concrete and easy to understand, related to TiDB is preferred.

Reviewer: ```suggestion
Many important topics have already been covered in the above topics, based on which course developers design the course content. The final content should be concrete and easy to understand, and, preferably, related to TiDB.
```

Reviewer: Why require translating into Chinese? I don't think it's necessary, as you are writing this in English.

### Programming Assignments

> For programming assignments, a distributed database system framework will be constructed. Even though there are many programming scaffolds about the database system and the distributed system, but they are not strongly connected, also C++ and Java are not easy to be mastered for young students. The new framework will be built with Golang, which is simpler and also the horsepower of the TiDB ecosystem. The content of the project will be borrowed from course materials listed below:

Reviewer: ```suggestion
For programming assignments, we will be constructing a distributed database system framework. There are many existing programming courses about database systems and distributed systems, but they are not strongly connected. The languages they use, C++ or Java, are not friendly to young students. Therefore, we use Golang to build the framework, which is more students friendly and also the primary language used in the TiDB ecosystem. The content of the project will be borrowed from course materials listed below:
```

- [CMU Bustub](https://github.com/cmu-db/bustub)
- [UCB moocbase](https://github.com/berkeley-cs186/fa19-moocbase)
- [MIT SimpleDB](https://github.com/MIT-DB-Class/simple-db-hw)
- [MIT SimpleKV](https://github.com/MIT-DB-Class/simpleKV)

## Rationale

> For TiDB Community, an educational distributed database system project in Golang and readings for novice are necessary, they are invitation cards of welcome young people to TiDB. We can open the door of the campus, and push computer system course education in universities forward.

Reviewer: ```suggestion
For TiDB Community, an educational distributed database system project in Golang for the novice is necessary. With  friendly and organized course content as invitation cards for young people to TiDB, we can open the door wider to campus, and push forward computer system course education in universities.
```

Reviewer: Can you explain what is "computer system course education "

Author: I was thinking about "education of computer systems", I think this can be deleted, it's verbose.

## Compatibility

TiDB will not be affected by this project.

## Implementation

To implement it, we need the following main steps:

> - Discussion the topics must be covered in reading materials, and contents of programming assignments.

Reviewer: ```suggestion
- Discuss the topics to be covered in reading materials and programming assignments design.
```

> - Build the architecture of the new TiDB for education.Experienced engineers from TiDB and TiKV are significant help.

Reviewer: > the architecture of the new TiDB for education

Not sure what this means here

> - Divide the course contents into multiple topics, each one will be handled by course development participants.

Reviewer: ```suggestion
- Divide the course topics, and assign them to course developers.
```

# Proposal: Create a distributed database system course for PingCAP University

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2019-12-6

> - Discussion at:

Reviewer 2: ```suggestion
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)
```
